`lib/uksched`: New scheduling API

skuenzer opened the pull request:
This PR re-factors the `lib/uksched`-API. Main purpose is the API simplifacation and removing features that aren't needed and would be covered by thread API libraries (like `pthread-embedded`, `musl`). The updated API is strictly separating thread creation from scheduling: threads can be created in various ways in terms of allocation ownership. Afterwards, they need to be assigned to a scheduler instance for getting executed. The idea of the API is also that any thread can also be unassigned and re-assigned to a different instance.

This PR currently does not support blocking thread garbage collection callbacks and blocking thread destructor functions. `lib/ukschedcoop`'s idle thread would not be non-blocking anymore which would break the scheduler implementation.

GitHub-Depends-On: #535

skuenzer committed:
lib/uksched: Re-worked thread creation interface

This commit introduces a new thread creation interface design.
Focus is on supporting the various conditions under which threads
may be created. Especially the different allocation ownership
models are supported. A thread created by `uksched` can allocate
and manage all resources by itself (stack, TLS, TCB, ectx) or
can just use given allocations.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/ukschedcoop: Re-factor scheduler instance handling

This commit re-factors the scheduler creation call by renaming if from
`uk_schedcoop_init()` to `uk_schedcoop_create()` because the scheduler
is allocating an instance object. Additionally, the memory layout is
simplified by embedding `struct uk_sched` to `struct uksched_coop`.
The `struct uk_sched_coop` is treated privately to the scheduler
implementation. The allocator that should be utilized by the scheduler
for allocation is also stored on the generic `struct uk_sched`.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/ukblkdev, lib/uknetdev, plat/xen: Thread functions are non-returning

As the re-factored uksched API requires, every threading functions
have to be non-returning. A thread function should call
`uk_sched_exit()` in order to terminate.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/ukschedcoop: Halt CPU from idle thread

We halt the current logical CPU within the idle thread so that we can
later compute the CPU utilization by simply measuring the scheduled
CPU time of each thread. The CPU utilization of a thread is its ratio
to the idle thread.

Checkpatch-Ignore: VOLATILE
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/ukdebug: Option to print thread name along with message

This commit introduces the option to print the thread name or, if
unnamed, the address of the TCB (struct uk_thread) of the currently
scheduled thread along with a debug or kernel log message. in case
`uk_thread_current()` returns NULL, the placeholder "<<n/a>>" is
printed instead.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Remove uneccessary UK_THREADF_QUEUEABLE

The flag UK_THREADF_QUEUEABLE is not needed anymore by the re-factored
scheduling API.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Thread garbage collection

Even terminated threads cannot simply release (e.g. stack, `struct
uk_thread`) their own resources. For this purpose, we hook such threads
into a scheduler-implementation-agnostic garbage collection list, which
is cleaned up by calling `uk_sched_thread_gc()`. The scheduler
implementation can choose the time at which this function should be
called.
In this commit, we also rename `uk_sched_thread_kill()` to
`uk_sched_thread_terminate()` to reduce possible confusion with the
handling of signals and to be consistent with the naming scheme of
`inittab` and `termtab`.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/ukschedcoop: `run_queue` and `sleep_queue`

For simplifying the code, this commit renames `thread_list` to
`run_queue` (threads that are ready to run) and `sleeping_threads` to
`sleep_queue` (threads that are sleeping and waiting to be woken up).

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: List of assigned threads per scheduler instance

A list of assigned threads is maintained with the general uksched API
for each scheduler instance. This allows implementatiobs to iterate
over this list and to collect for thread information.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/ukschedcoop: Remove indirection for `schedcoop_schedule()`

Removes the unecesary indirection for `schedcoop_schedule()`.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Re-factor `uk_sched_thread_create()` helper

The `uk_sched_thread_create()` is re-factored to be more consistent with
the `uk_thread_create_*()` API. The references to thread attributes are
removed as they are not in use.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/ukschedcoop: Create idle thread within scheduler implementation

The idle thread is created as part of a scheduler implementation.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/ukschedcoop: Call garbage collection as part of idle thread

Calls the thread garbage collection as part of the idle thread in
order to give it lower priority and to speed up scheduling the next
thread.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Remove unused thread_attr.c|.h

Thread attributes are currently unused and are therefore removed. It is
intended that POSIX-like thread attributes are implemented with a POSIX
thread API abstraction library.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Remove thread waiting and detached threads

We envision that thread waiting and detached threads will be
be implemented by a POSIX thread API abstraction library. The idea is
to have a minimal `uksched` implementation that avoids unnecessary
functionality.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/ukboot: Allocate TLS for boot thread

Allocates memory to the boot context for thread-local storage when
`ukalloc` is part of the build. This is for two reasons:
1) If scheduling is not enabled, code that nevertheless refers to
   on a TLS variables would still work.
2) We plan to continue to use the boot context as an init thread
   when scheduling is initiated.
In addition, we stop booting if the heap could not be initialized,
because then we also cannot allocate a TLS.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Introduce `uk_thread_uktls_var()`

This commit introduces the ability to modify TLS variables of an
foreign thread. The TLS is used as an extensible way to implement
Unikraft's TCB. Such an access method is required to allow
initialization of foreign threads. The implemented method works
only with static TLSes for now.
In addition, changing the currently active TLS temporarily is
supported as well: On thread switches, the CPU register containing
the TLS pointer is saved and restored. For this purpose, the
reference of the original TLS is stored in the `uktlsp` field
within `uk_thread`.
This mode can also be utilized for binary compatible modes where
externally built binaries are referring to a different TLS than
code that was built with Unikraft.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Remove reference to newlib (`_reent`)

Entries to the TCB should be done with the TLS instead. This leads to
a cleaner and independent separation of the code in libraries.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Provide thread container creation

There are use cases where a thread needs to be created before
it can be initialized with a context. For this purpose,
this commit introduces `uk_thread_create_container()` and
`uk_thread_create_container2()`. The first function takes care
of allocating a stack and TLS, while the second variant expects
a stack and TLS to be given by the caller.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Replace `uk_sched_get_default()` with `uk_sched_current()`

Instead of maintaining a default scheduler this commit provides a
function that returns the current scheduler of the current thread.
The original intention was to have a fall-back scheduler to which a
newly created thread should be assigned to. Using the current scheduler
as the fall-back policy is more appropriate when multiple schedulers are
used in the system.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Option to enable debug messages

This commit introduces a configuration option that enables all
debug messages from `uksched`.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Custom TCB initialization hooks

This commit introduces the possibility for a third-party threading
library to initialize a custom thread control block (TCB). The layout
of the TLS and TCB is defined by the compiler of the target architecture
and of which Unikraft only utilizes the TLS. As soon as a library
(e.g., pthread-embedded, musl) activates the invisible configuration
option `CONFIG_LIBUKSCHED_TCB_INIT`, the following two
hooks are called during thread creation and removal:
- `uk_thread_uktcb_init()`: On TLS allocation during thread creation
- `uk_thread_uktcb_fini()`: On TLS removal of a thread
These hooks are called by any `lib/uksched` API call that create or
terminate a thread so that the uksched API seamlessly integrates into
the third party threading library.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Thread creation helper functions

This commit introduces thread creation helper functions that create a
thread that is directly assigned to the current scheduler. For the
creation default allocators are used that are set by the scheduler.
Variants for the most common cases are provided:
- `uk_sched_thread_create_fn0()`: thread function without any argument
- `uk_sched_thread_create_fn1()`: thread function with one argument
- `uk_sched_thread_create_fn2()`: thread function with two arguments
In addition, the shortcut macro `uk_sched_thread_create()` is provided
for a convenient thread creation with just four arguments: (1) the
scheduler instance, (2) the thread function to execute (one argument),
(3) an argument pointer, and (4) an optional thread name.
This macro basically replaces the previous helper `uk_thread_create`()`.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Remove unused `threads_started` from `struct uk_sched`

Because the field `threads_started` has no use at the moment,
we remove it from `struct uk_sched` to simplify the API.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Re-factor thread state flags helper

The helpers for querying and setting the thread state are refactored
to fit the naming scheme and to avoid name clashing.

Checkpatch-Ignore: SINGLE_STATEMENT_DO_WHILE_MACRO
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Call termination callbacks on thread exit

The thread termination callbacks were called only when a thread was
released. Especially for self-terminating threads this meant that the
termination callback was not called until the garbage collector
released the thread resources.
This commit implements the expected behavior, namely that the thread
termination callbacks are called when the thread transitions to the
exited state. Whenever `uk_thread_release()` is called on a non-exited
thread, the state transition to the exited state is enforced first. This
ensures that termination callbacks are called even in such a case.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Provide GC hook with thread exit (`uk_sched_thead_exit2()`)

The `uk_sched_thread_exit2()` function behaves like
`uk_sched_thread_exit()` (self-exiting thread) but provides the ability
to execute a callback function when the thread context is released by
the garbage collector. This can be useful for libraries like `musl`
which need to be able to release library managed resources (like TLS
and stack) on a higher level thread exit interface
(e.g. `pthread_exit`).

Please note that this self-exit variant is currently a workaround that
is used by musl. A more generic implementation could be done by
maintaining a list of GC hooks per thread and call each entry during
garbage collection.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

dragosargint reviewed (approved):
The PR looks good! Here are some tests that I used for this PR [1]
[1] https://github.com/dragosargint/test

adinasm commented on lib/posix-futex/futex.c:
```diff
@@ -178,7 +178,7 @@ static int futex_wake(uint32_t *uaddr, uint32_t val)
 			/* TODO: Replace with uk_thread_wakeup when the new
```
This comment is no longer needed.

adinasm commented on lib/uksched/include/uk/sched.h:
```diff
@@ -115,160 +109,200 @@ static inline void uk_sched_yield(void)
 
 void uk_sched_thread_sleep(__nsec nsec);
+
+/* exits the current thread context */
```
```suggestion
/* Exits the current thread context */
```

adinasm commented on lib/uksched/include/uk/sched.h:
```diff
@@ -115,160 +109,200 @@ static inline void uk_sched_yield(void)
+/* exits the current thread context */
 void uk_sched_thread_exit(void) __noreturn;
 
+/* exits the current thread and runs the given routine when releasing the
```
```suggestion
/* Exits the current thread and runs the given routine when releasing the
```

adinasm commented on lib/uksched/include/uk/sched.h:
```diff
@@ -115,160 +109,200 @@ static inline void uk_sched_yield(void)
+ */
+void uk_sched_thread_exit2(uk_thread_gc_t gc_fn, void *gc_argp) __noreturn;
+
+/* terminate another thread */
```
```suggestion
/* Terminates another thread */
```

adinasm commented on lib/uksched/include/uk/tcb_impl.h:
```diff
@@ -0,0 +1,96 @@
+ * It is intended that a libc or a thread abstraction library that require
+ * a TCB will provide this function symbol.
+ *
+ * @param uktcb Reference to reserved space to custom thread control block
```
```suggestion
 * @param tcb Reference to reserved space to custom thread control block
```

adinasm reviewed (commented):
Looks good to me, I only left some minor comments.

skuenzer committed:
lib/uksched: `uk_sched_start()` returns

This commit changes `uk_sched_start()` so that the resources
of the current context (stack, TLS) continue to execute under
the started scheduler as `init` thread.
Because the thread resources (stack, TLS) are managed external
to the started scheduler, this gives the `init` thread the
opportunity to cleanly free up its resources on exit and
may even enable to start scheduling as part of the `inittab`.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Refactor `uk_sched_thread_(blocked|woken)()`

In order to simplify the API and to avoid implementation mistakes,
`uk_sched_thread_woken()` and `uk_sched_thread_blocked()` require only
the thread as argument. The corresponding scheduler is retrieved from
the given thread object instead.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Remove references to `lib/uksignal`

The implementation of `uksignal` is currently broken and needs to be
re-done. Additionally, the TLS should be used instead by libraries to
make entries to the TCB. This leads to a cleaner and independent
separation of the code in libraries.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/ukboot, lib/uksched: Initialize schedulers similar to allocators

As an initial preparation for SMP and for consistency, initializing a
scheduler is done in the same way as selecting and initializing a heap
allocator with the bootstrapping library. This scheme gives us full
flexibility for custom scheduler layout initializations.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/uksched: Header for scheduler implementations

Creates a new header `<uk/sched_impl.h>` that should only be used by
actual scheduler implementations. Corresponding function declarations
and macros are moved from `<uk/sched.h>` to this new header.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

craciunoiuc commented:
Hey @skuenzer

Please check the checkpatch results + CodeFactor + Codacy and fix what's fixable

skuenzer commented:
> Hey @skuenzer
> 
> Please check the checkpatch results + CodeFactor + Codacy and fix what's fixable

I am already on it 😉

skuenzer commented:
Hi, thanks a lot for all the comments, I updated this PR and applied your suggestions. I saw that Codacy is reporting 4 issues but IMO these are false positives (becuase of nested macros and a macro that expands to a for-loop).

razvand reviewed (commented):
This is OK with me. I'll wait for comments from @adinasm, @dragosargint and @craciunoiuc and then mark it as approved and add the reviewed and approved tags.

@skuenzer , as with PR #535 , I don't know what causes [`UK_ASSERT` Codacy issues](https://github.com/unikraft/unikraft/pull/564/checks).

razvand commented on lib/ukschedcoop/schedcoop.c:
```diff
@@ -70,186 +71,212 @@ static void schedcoop_schedule(struct uk_sched *s)
 				uk_thread_wake(thread);
-
-			else if (thread->wakeup_time < min_wakeup_time)
+			else if (!min_wakeup_time
```
This Codacy warning should be ignores. `min_wakeup_time` is `0` only the first time. Then it will be set to `thread->wakeup_time`.

dragosargint reviewed (approved):
I retested this PR and apart from the comments left by @razvand, I found no other problems.
Let it shine!
Reviewed-by: Dragos Iulian Argint <dragosargint21@gmail.com>

craciunoiuc commented on plat/common/arm/thread_start64.S:
```diff
@@ -29,7 +29,7 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
-#include <uk/plat/common/sw_ctx.h>
+//#include <uk/plat/common/ctx.h>
```
Why is this commented and not just removed?

Will it be used in the future?

skuenzer replied:
I agree, this is weird. I removed it. I did this as part of PR #535. Because this PR depends on #535, you are seeing its commits also here.

craciunoiuc commented on arch/arm/arm64/ctx.S:
```diff
@@ -0,0 +1,162 @@
+ *          Robert Kuban <robert.kuban@opensynergy.com>
+ *
+ * Copyright (c) 2018, Arm Ltd. All rights reserved.
+ * Copyright (c) 2022, OpenSynergy GmbH All rights reserved.
```
```suggestion
 * Copyright (c) 2022, OpenSynergy GmbH. All rights reserved.
```

skuenzer replied:
Thanks for spotting this :+1:

skuenzer replied:
I fixed it with PR #535 .

craciunoiuc commented on lib/uksched/include/uk/sched.h:
```diff
@@ -1,8 +1,12 @@
+ *          Simon Kuenzer <simon@unikraft.io>
  *
  * Copyright (c) 2017, NEC Europe Ltd., NEC Corporation. All rights reserved.
+ * Copyright (c) 2022, NEC Laboratories GmbH, NEC Corrporation,
+ *                     All rights reserved.
```
```suggestion
 * Copyright (c) 2022, NEC Laboratories GmbH, NEC Corrporation.
 *                     All rights reserved.
```

skuenzer replied:
Thanks and fixed :+1:

craciunoiuc commented on lib/uksched/thread.c:
```diff
@@ -101,107 +87,747 @@ extern const struct uk_thread_inittab_entry _uk_thread_inittab_end;
+	return _uk_thread_call_inittab(t);
+}
+
+int uk_thread_init_bare_fn0(struct uk_thread *t,
+			    uk_thread_fn0_t fn,
+			    uintptr_t sp,
+			    uintptr_t tlsp,
+			    bool is_uktls,
+			    struct ukarch_ectx *ectx,
+			    const char *name,
+			    void *priv,
+			    uk_thread_dtor_t dtor)
```
Is there a convention to how function arguments should be structured?

For example, I added a newline only when passing 80 characters.

skuenzer replied:
Yes, this is true. I do this normally, too. In fact, I have a kind of exception only with functions with a lot of arguments (don't ask me for the threshold that I am using 😉 ). I noticed that pressing as much as arguments possible into a few lines does not contribute to the readability. Here, we actually have a rather unbalanced length of each argument that I find tedious to read. I am prefer keeping it this way because of this, what do you think?

craciunoiuc replied:
Yeah, I also like it like this more. Just wanted to know if there's a convention for this, for the future.

I'm marking as resolved.

craciunoiuc commented on lib/ukschedcoop/Config.uk:
```diff
@@ -1,4 +1,5 @@
 	default y
-	depends on LIBUKSCHED
+	select LIBNOLIBC if !HAVE_LIBC
+	select LIBUKSCHED
```
```suggestion
	select LIBUKSCHED

```

craciunoiuc replied:
Should there be a newline here?

skuenzer replied:
Yes, an ending newline should be there. I updated my editor settings now. The newline was missing originally but should be there. I fixed it :+1:

craciunoiuc reviewed (changes requested):
Hey @skuenzer,

Some final touches from my side, I will approve after this.

`lib/ukschedcoop: Re-factor scheduler instance creation and instance object` -> This commit message is a bit too long unfortunately 😢
`lib/uksched: Call termination callbacks when thread transitions to exited` -> also this one
`lib/uksched: Provide GC hook on thread exit with uk_sched_thead_exit2()` -> also this one
`lib/ukboot, lib/uksched: Scheduler initialization inline with lib/ukalloc` -> also this one

skuenzer commented:
> Hey @skuenzer,
> 
> Some final touches from my side, I will approve after this.
> 
> `lib/ukschedcoop: Re-factor scheduler instance creation and instance object` -> This commit message is a bit too long unfortunately 😢 `lib/uksched: Call termination callbacks when thread transitions to exited` -> also this one `lib/uksched: Provide GC hook on thread exit with uk_sched_thead_exit2()` -> also this one `lib/ukboot, lib/uksched: Scheduler initialization inline with lib/ukalloc` -> also this one

Hey @craciunoiuc , I updated the PR (including shorter commit messages, yes it is better 😉 ). Please, have a look. I prefered keeping the argument list with `uk_thread_init_bare_fn0()` and co, the rest I fixed. Thanks a lot for your review.

craciunoiuc reviewed (approved):
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

adinasm reviewed (approved):
This is great. Thanks!
Reviewed-by: Adina Smeu <adina.smeu@gmail.com>

razvand commented:
@skuenzer , the merging of #535  is now requiring a rebase of this PR. This mostly means getting rid of all commits belonging to #535 and then seeing if there are remaining conflicts.

skuenzer commented:
> @skuenzer , the merging of #535 is now requiring a rebase of this PR. This mostly means getting rid of all commits belonging to #535 and then seeing if there are remaining conflicts.

Sure, I just pushed the rebased version ✅ .
It was conflict-free because this PR is a branch that was based on the branch of PR #535 (the one that just got merged)... 🤔 Unfortunately, GitHub could not detect that commits of this PR were already integrated. I am not sure if there is a different way how we could handle these "stacked PRs" with GitHub... 🤷🏻‍♂️ but it seems we rely on this manual step after a merge.

razvand reviewed (approved):
Thanks, @skuenzer 

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
